Add volume bar per station

GeheimagentNr1 opened the pull request:
Add volume bar per station, so you can listen to different station, without the problem to not hear the own frequency

The volume gains are saved to the local storage with the station callsign, but are set to the backend using the frequency, using a mechanism already existing for unicom and guard. That should be good, because the probability of trying to listen to several stations with the same frequency is probably relatively low.

![image](https://github.com/user-attachments/assets/596a8f49-f47a-4edc-bfcf-c7c8f09addf9)

GeheimagentNr1 commented:
After a discussion with @neilenns, I will add a configuration flag "Enable per-station volume", which sets if the station volume bars are shown or not.

GeorgeBarlow commented:
Great idea! However, I'm not sure of the design or how often people would need to adjust the volume of every single radio on their radio list and I'm worried that this could give quite the cluttered look if every radio shows a volume bar.

I would argue a more unobtrusive approach would suit the look and feel of the radio container much better. An example of what I mean would be something like this:

https://github.com/user-attachments/assets/dda3b93a-68af-428c-8307-e886b4a5fab1

Do bare in mind that the example was merely put together in 5 or so minutes therefore the UX and other things are to be desired. 

It would be great to discuss this further so let me know your thoughts, once again I can see the benefit of this it could be a great feature to add :)

GeheimagentNr1 commented:
> Great idea! However, I'm not sure of the design or how often people would need to adjust the volume of every single radio on their radio list and I'm worried that this could give quite the cluttered look if every radio shows a volume bar.
> 
> I would argue a more unobtrusive approach would suit the look and feel of the radio container much better. An example of what I mean would be something like this:
> 
> 3Y7H3y.mp4  
> Do bear in mind that the video above is only an example and was merely put together in 5 or so minutes therefore the UX and other things are to be desired (+ also the headphone icon placement would potentially clash with the incoming HF UI improvement)
> 
> It would be great to discuss this further so let me know your thoughts, once again I can see the benefit of this it could be a great feature to add :)

That looks great, but my CSS skills are not good enough to do that.  
About the icon: Maybe it would an idea to show the audio icon in the top left corner and move the HF icon below the audio icon.

GeorgeBarlow commented:
> > Great idea! However, I'm not sure of the design or how often people would need to adjust the volume of every single radio on their radio list and I'm worried that this could give quite the cluttered look if every radio shows a volume bar.
> > I would argue a more unobtrusive approach would suit the look and feel of the radio container much better. An example of what I mean would be something like this:
> > 3Y7H3y.mp4
> > Do bear in mind that the video above is only an example and was merely put together in 5 or so minutes therefore the UX and other things are to be desired (+ also the headphone icon placement would potentially clash with the incoming HF UI improvement)
> > It would be great to discuss this further so let me know your thoughts, once again I can see the benefit of this it could be a great feature to add :)
> 
> That looks great, but my CSS skills are not good enough to do that. If you can tell me, how to achieve this, it would be appreciated. About the icon: Maybe it would an idea to show the audio icon in the top left corner and move the HF icon below the audio icon.

What I've done is I've pushed the example and a few functional extras to a branch on TrackAudio: 
 https://github.com/pierr3/TrackAudio/tree/hf%2Bindividual-volume
 
 You can take inspiration from it if you like, it currently depends on HF, the reason was because I wanted to see how it would look with that. Feel free to diverge from it or improve on it :) 

For reference this is what it looks like:
https://github.com/user-attachments/assets/b855fbf4-97ae-455b-b57c-c58c2d658963

GeheimagentNr1 commented:
> > > Great idea! However, I'm not sure of the design or how often people would need to adjust the volume of every single radio on their radio list and I'm worried that this could give quite the cluttered look if every radio shows a volume bar.
> > > I would argue a more unobtrusive approach would suit the look and feel of the radio container much better. An example of what I mean would be something like this:
> > > 3Y7H3y.mp4
> > > Do bear in mind that the video above is only an example and was merely put together in 5 or so minutes therefore the UX and other things are to be desired (+ also the headphone icon placement would potentially clash with the incoming HF UI improvement)
> > > It would be great to discuss this further so let me know your thoughts, once again I can see the benefit of this it could be a great feature to add :)
> > 
> > 
> > That looks great, but my CSS skills are not good enough to do that. If you can tell me, how to achieve this, it would be appreciated. About the icon: Maybe it would an idea to show the audio icon in the top left corner and move the HF icon below the audio icon.
> 
> What I've done is I've pushed the example and a few functional extras to a branch on TrackAudio: https://github.com/pierr3/TrackAudio/tree/hf%2Bindividual-volume
> 
> You can take inspiration from it if you like, it currently depends on HF, the reason was because I wanted to see how it would look with that. Feel free to diverge from it or improve on it, once we complete the feature we'll make sure that we push it through this PR so you get credited as I copied your code into that branch and don't want it to look like it was under my commit :)
> 
> For reference this is what it looks like:
> 
>  6p5V6k.mp4

Thank you.
I'm on it.

Would it be a good Idea to add a reset button for the volume per station?
Because if a user logs in as EDDH_TWR, EDDH_E_GND or EDDH_DEL different volumes might be required.
Because after a station change, the main frequency having a low volume might be bad.

Multiple ideas:
- Add a reset button for the volume per station values?
- Reset the volume per station on restart?
   - might be bad, if TrackAudio crashed. Or you observe multiple days in the same stations.
- Reset the volume per station, if the log in callsign changes.
   - might be bad, if you change from EDDH_DEL to EDDH_TWR and the setting for EDDH_E_GND should stay the same.
- Reset the volume per station, if TX is activated, so the relevant/station on which you are talking are always load enough.
   - Alternative, force the volume for station for which TX is activated always to 100, without resetting the volume per station setting. With this idea, it would be possible to set lowered volumes for and listen frequencies and force the relavent/talking frequencies to 100 percent.
   
   What do you think?
   I think the force volume for stations with TX activated to 100%, would be the best solution,

neilenns commented:
How about double click on the slider resets the volume? That gives the user control over it but doesn't add more elements to the UI.

neilenns commented:
Can this be exposed via the websocket interface too? That way it could be controlled via stream deck, I can make a dial action for it or can auto set the volume to a specific level when connected.

GeheimagentNr1 commented:
> Can this be exposed via the websocket interface too? That way it could be controlled via stream deck, I can make a dial action for it or can auto set the volume to a specific level when connected.
> 
> To send the value to clients include a property called `volume` with the current setting here: https://github.dev/pierr3/TrackAudio/blob/d0483b0cc779d4cac7943aa9030cf59a1812aba8/backend/src/sdk.cpp#L34-L53
> 
> To process the request, look for a `volume` property in the incoming message and use it here: https://github.dev/pierr3/TrackAudio/blob/d0483b0cc779d4cac7943aa9030cf59a1812aba8/backend/src/sdk.cpp#L256-L257

Are changes to the UI part necessary too?  
Because I don't know if it would be shown in the UI, if it's changed via the websocket.

GeorgeBarlow commented:
I've just made some improvements to the branch, had to fix some stuff in the backend as it was doing some weird stuff with re-updating the gain when we didn't want it to. Check it out here's a video that kinda goes over some key changes. 


https://github.com/user-attachments/assets/38be05da-0353-4030-ada4-1c8ec26f0e53

I also added the TX thing, so when you click TX it sets the volume of that frequency to the MAIN gain if it is higher than what the current value of that radio gain is, if it was previously in manual mode.

neilenns commented:
The websocket would just be as separate method of setting it. Presumably if everything is wired up correctly the UI would update to match whatever value was sent from a stream deck to set the gain.

neilenns commented:
> I've just made some improvements to the branch, had to fix some stuff in the backend as it was doing some weird stuff with re-updating the gain when we didn't want it to. Check it out here's a video that kinda goes over some key changes. 
> 
> 
> https://github.com/user-attachments/assets/38be05da-0353-4030-ada4-1c8ec26f0e53
> 
> I also added the TX thing, so when you click TX it sets the volume of that frequency to the MAIN gain if it is higher than what the current value of that radio gain is, if it was previously in manual mode.

Seems like there is enough logic here that I suggest leaving my websocket idea to a separate PR. Otherwise this will just drag on for a long time. I can add web socket support sometime down the road.

GeorgeBarlow commented:
> The websocket would just be as separate method of setting it. Presumably if everything is wired up correctly the UI would update to match whatever value was sent from a stream deck to set the gain.

We would either have to manage the state of the volume set inside the C++ side of TrackAudio or update AFV Native to extract the current volume set out, both of these probably would require a separate PR as it involves changing a few things. So yeh maybe open a separate one once this gets out.

GeheimagentNr1 commented:
It still needs to be checked, if after a change of the Main Volume, if the Frequency specific volumes are still in place.

Because of SetRadioGain and SetFrequencyRadioGain methods in the main.cpp, both setting volumes per frequency to the afvClient.  
  
Before I thought, that the volume of the frequency and the main volume is set.  
So if the main volume is set to 50 and the main volume is set to 50, that the output volume is 25. So 50% of the main volume.  
That would make more sense, because if you want to volume down all frequencies, and you have set for multiple frequencies a custom volume, the volume of ever frequency with custom volume needs to be changed.

In the current implementation the volume is changed, if the TX button is activated or deactivated.  
In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume percentage of the main volume, if a custom volume is set.

GeheimagentNr1 commented:
Is there is a way to test TX activation without connecting as controller to vatsim?

GeorgeBarlow commented:
> It still needs to be checked, if after a change of the Main Volume, if the Frequency specific volumes are still in place.
> 
> Because of SetRadioGain and SetFrequencyRadioGain methods in the main.cpp, both setting volumes per frequency to the afvClient.
> 
> Before I thought, that the volume of the frequency and the main volume is set. So if the main volume is set to 50 and the main volume is set to 50, that the output volume is 25. So 50% of the main volume. That would make more sense, because if you want to volume down all frequencies, and you have set for multiple frequencies a custom volume, the volume of ever frequency with custom volume needs to be changed. Even with the guard and unicom volume, I always thought, that it will be 50% of the main volume.
> 
> In the current implementation, the volume is not changed, if the TX button is activated or deactivated/clicked. In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume percentage of the main volume, if a custom volume is set.

The only difference between setRadioGain and setRadioGainAll is just that setRadioGainAll performs a for loop over all the known radio states and sets the gain to that one value. @pierr3 correct me if I'm wrong but that's what I can see on the afv_native side?

```ts
  export function SetRadioGain(gain: number): void;
  export function SetFrequencyRadioGain(frequency: number, gain: number): void;
  ```
  
 These are the only two exported functions from the C++ side that interact with the gain, where the first just sets that gain for all radios and the second sets it for a specific radio.

>In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume >percentage of the main volume, if a custom volume is set.

Well in that case just update the `useEffect` to something like:
```ts
 const setStoredGain = () => {
    const storedGain = window.localStorage.getItem(radio.callsign + 'RadioGain');
    const gainToSet = storedGain?.length ? parseInt(storedGain) : radioGain;
    if (storedGain) {
      updateRadioGainValue(gainToSet, false, false);
    }
  }


  useEffect(() => {
    if (radio.tx) {
      updateRadioGainValue(100, true, false);
    } else {
      setStoredGain();
    }
  }, [radio.tx]);

```

Untested of course but you get the idea

GeheimagentNr1 commented:
> > It still needs to be checked, if after a change of the Main Volume, if the Frequency specific volumes are still in place.  
> > Because of SetRadioGain and SetFrequencyRadioGain methods in the main.cpp, both setting volumes per frequency to the afvClient.  
> > Before I thought, that the volume of the frequency and the main volume is set. So if the main volume is set to 50 and the main volume is set to 50, that the output volume is 25. So 50% of the main volume. That would make more sense, because if you want to volume down all frequencies, and you have set for multiple frequencies a custom volume, the volume of ever frequency with custom volume needs to be changed. Even with the guard and unicom volume, I always thought, that it will be 50% of the main volume.  
> > In the current implementation, the volume is not changed, if the TX button is activated or deactivated/clicked. In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume percentage of the main volume, if a custom volume is set.
> 
> The only difference between setRadioGain and setRadioGainAll is just that setRadioGainAll performs a for loop over all the known radio states and sets the gain to that one value. @pierr3 correct me if I'm wrong but that's what I can see on the afv\_native side?
> 
> ```ts
>   export function SetRadioGain(gain: number): void;
>   export function SetFrequencyRadioGain(frequency: number, gain: number): void;
> ```
> 
> These are the only two exported functions from the C++ side that interact with the gain, where the first just sets that gain for all radios and the second sets it for a specific radio.
> 
> > In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume >percentage of the main volume, if a custom volume is set.
> 
> Well in that case just update the `useEffect` to something like:
> 
> ```ts
> const setStoredGain = () => {
>     const storedGain = window.localStorage.getItem(radio.callsign + 'RadioGain');
>     if (storedGain) {
>       const gainValue = parseInt(storedGain);
>       if (isNaN(gainValue)) {
>         resetToMainGain();
>       } else {
>         updateRadioGainValue(gainValue, true, false);
>       }
>     } else {
>       resetToMainGain();
>     }
>   };
> 
>   useEffect(() => {
>     if (radio.tx) {
>       updateRadioGainValue(100, true, false);
>     } else {
>       setStoredGain();
>     }
>   }, [radio.tx]);
> ```
> 
> Untested of course but you get the idea, though since MAIN doesn't actually control the main overall audio leaving TrackAudio, we either reduce the other volume gains by MAIN mathematically before setting it or we make this feature of doing 100 optional.

I see. I did the useEffect thing, in another method and in the wrong place.  
  
I will think about, how Frequency and Main volume could work together in the next few days.  
For now, I changed the PR to a Draft.

GeorgeBarlow commented:
> > > It still needs to be checked, if after a change of the Main Volume, if the Frequency specific volumes are still in place.
> > > Because of SetRadioGain and SetFrequencyRadioGain methods in the main.cpp, both setting volumes per frequency to the afvClient.
> > > Before I thought, that the volume of the frequency and the main volume is set. So if the main volume is set to 50 and the main volume is set to 50, that the output volume is 25. So 50% of the main volume. That would make more sense, because if you want to volume down all frequencies, and you have set for multiple frequencies a custom volume, the volume of ever frequency with custom volume needs to be changed. Even with the guard and unicom volume, I always thought, that it will be 50% of the main volume.
> > > In the current implementation, the volume is not changed, if the TX button is activated or deactivated/clicked. In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume percentage of the main volume, if a custom volume is set.
> > 
> > 
> > The only difference between setRadioGain and setRadioGainAll is just that setRadioGainAll performs a for loop over all the known radio states and sets the gain to that one value. @pierr3 correct me if I'm wrong but that's what I can see on the afv_native side?
> > ```ts
> >   export function SetRadioGain(gain: number): void;
> >   export function SetFrequencyRadioGain(frequency: number, gain: number): void;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > These are the only two exported functions from the C++ side that interact with the gain, where the first just sets that gain for all radios and the second sets it for a specific radio.
> > > In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume >percentage of the main volume, if a custom volume is set.
> > 
> > 
> > Well in that case just update the `useEffect` to something like:
> > ```ts
> > const setStoredGain = () => {
> >     const storedGain = window.localStorage.getItem(radio.callsign + 'RadioGain');
> >     if (storedGain) {
> >       const gainValue = parseInt(storedGain);
> >       if (isNaN(gainValue)) {
> >         resetToMainGain();
> >       } else {
> >         updateRadioGainValue(gainValue, true, false);
> >       }
> >     } else {
> >       resetToMainGain();
> >     }
> >   };
> > 
> >   useEffect(() => {
> >     if (radio.tx) {
> >       updateRadioGainValue(100, true, false);
> >     } else {
> >       setStoredGain();
> >     }
> >   }, [radio.tx]);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Untested of course but you get the idea, though since MAIN doesn't actually control the main overall audio leaving TrackAudio, we either reduce the other volume gains by MAIN mathematically before setting it or we make this feature of doing 100 optional.
> 
> I see. I did the useEffect thing, in another method and in the wrong place.
> 
> I will think about, how Frequency and Main volume could work together in the next few days. For now, I changed the PR to a Draft.

No problem, let me know how you get on. I've put an implementation of the individual radio gains being relative to the MAIN gain in the branch let me know if that is what you mean, though gain has a very weird order of magnitude I've found so not sure how effective this is.

GeheimagentNr1 commented:
> > > > It still needs to be checked, if after a change of the Main Volume, if the Frequency specific volumes are still in place.  
> > > > Because of SetRadioGain and SetFrequencyRadioGain methods in the main.cpp, both setting volumes per frequency to the afvClient.  
> > > > Before I thought, that the volume of the frequency and the main volume is set. So if the main volume is set to 50 and the main volume is set to 50, that the output volume is 25. So 50% of the main volume. That would make more sense, because if you want to volume down all frequencies, and you have set for multiple frequencies a custom volume, the volume of ever frequency with custom volume needs to be changed. Even with the guard and unicom volume, I always thought, that it will be 50% of the main volume.  
> > > > In the current implementation, the volume is not changed, if the TX button is activated or deactivated/clicked. In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume percentage of the main volume, if a custom volume is set.
> > > 
> > > The only difference between setRadioGain and setRadioGainAll is just that setRadioGainAll performs a for loop over all the known radio states and sets the gain to that one value. @pierr3 correct me if I'm wrong but that's what I can see on the afv\_native side?
> > > 
> > > ```ts
> > >   export function SetRadioGain(gain: number): void;
> > >   export function SetFrequencyRadioGain(frequency: number, gain: number): void;
> > > ```
> > > 
> > > These are the only two exported functions from the C++ side that interact with the gain, where the first just sets that gain for all radios and the second sets it for a specific radio.
> > > 
> > > > In the activation case it should be set to 100% of the main volume and in the deactivation case to the custom volume >percentage of the main volume, if a custom volume is set.
> > > 
> > > Well in that case just update the `useEffect` to something like:
> > > 
> > > ```ts
> > > const setStoredGain = () => {
> > >     const storedGain = window.localStorage.getItem(radio.callsign + 'RadioGain');
> > >     if (storedGain) {
> > >       const gainValue = parseInt(storedGain);
> > >       if (isNaN(gainValue)) {
> > >         resetToMainGain();
> > >       } else {
> > >         updateRadioGainValue(gainValue, true, false);
> > >       }
> > >     } else {
> > >       resetToMainGain();
> > >     }
> > >   };
> > > 
> > >   useEffect(() => {
> > >     if (radio.tx) {
> > >       updateRadioGainValue(100, true, false);
> > >     } else {
> > >       setStoredGain();
> > >     }
> > >   }, [radio.tx]);
> > > ```
> > > 
> > > Untested of course but you get the idea, though since MAIN doesn't actually control the main overall audio leaving TrackAudio, we either reduce the other volume gains by MAIN mathematically before setting it or we make this feature of doing 100 optional.
> > 
> > I see. I did the useEffect thing, in another method and in the wrong place.  
> > I will think about, how Frequency and Main volume could work together in the next few days. For now, I changed the PR to a Draft.
> 
> No problem, let me know how you get on. I've put an implementation of the individual radio gains being relative to the MAIN gain in the branch let me know if that is what you mean, though gain has a very weird order of magnitude I've found so not sure how effective this is.

Did you add the radioGain to the SetFrequencyState to prevent to have to send 2 requests for RX/RX/XC change and the volume change?

GeorgeBarlow commented:
> Did you add the radioGain to the SetFrequencyState to prevent to have to send 2 requests for RX/RX/XC change and the volume change?

It is because in the C++ side on state update we were reconfirming the gain and that was previously using the global gain stored in the C++ settings file, this could probably be removed from state update in all fairness though just for drafting purposes I left the ability to change it in there.

GeheimagentNr1 commented:
I created a new version of the feature from the bottom.

I renamed all variables in backend and frontend from gain to volume, because it is the more common name and values from 0 to 100 are used.
The calculation from volume (0-100) to gain (0.0-1.0) is now done in the RadioHelper in the backend.
Also, the combination of mainVolume and stationVolume to frequencyVolume/gain is now done on the backend of the RadioHelper.
When setting the main volume the frequencyVolume/gain is calculated from the new mainVolume and the stationVolume in the backend. Because of that the stationVolumes are now stored in the UserSession in the backend.

I added the volume bar with the settings icon to every station without a reset button.
The stationVolume is now force to 100 if TX is active for a station and the volume bar is disabled in this case.

GeorgeBarlow commented:
> I renamed all variables in backend and frontend from gain to volume, because it is the more common name and values from 0 to 100 are used. 

I get why you would rename it, volume does appear to describe it better, though I do think 'gain' does describes what is happening with that value better as in afv native we're directly multiplying the audio signal's amplitude using this value whereas volume is more the overall perceived loudness. Not only that, keeping it as gain provides better continuity throughout both TrackAudio and afv-native.

> Because of that, the stationVolumes are now stored in the UserSession in the backend.

Yeh that sounds good to have it all in one place, though we do need to find a way of either getting rid of the C++ config file and merging it (ideally) with the Electron one as we are managing two states now which is not really needed.

@pierr3 thoughts?

GeheimagentNr1 commented:
> > I renamed all variables in backend and frontend from gain to volume, because it is the more common name and values from 0 to 100 are used.
> 
> I get why you would rename it, volume does appear to describe it better, though I do think gain describes what is happening with that value better as in afv native we're directly multiplying the audio signal's amplitude using this value whereas volume is more the overall perceived loudness. Not only that, keeping it as gain provides better continuity throughout both TrackAudio and afv-native.

Because of that, I used volume for 0-100 values and gain for 0.00-1.00 values.

GeorgeBarlow commented:
> > > I renamed all variables in backend and frontend from gain to volume, because it is the more common name and values from 0 to 100 are used.
> > 
> > 
> > I get why you would rename it, volume does appear to describe it better, though I do think gain describes what is happening with that value better as in afv native we're directly multiplying the audio signal's amplitude using this value whereas volume is more the overall perceived loudness. Not only that, keeping it as gain provides better continuity throughout both TrackAudio and afv-native.
> 
> Because of that, I used volume for 0-100 values and gain for 0.00-1.00 values.

Oh apologies, didn't see that. I will have a proper skim through then. 👍

neilenns commented:
Catching up on this, a couple of comments:

1. I will eventually wind up opening a request for an "isMuted" state, so stations can have their audio turned off quickly. I plan to drive that from the WebSocket interface with Stream Deck buttons and dials, and will want a `true`, `false` and `"toggle"` option. The UI would, presumably eventually, also want to indicate a station is muted.

2. I will also wind up opening a request to set, increment, and decrement the volume of individual stations via the WebSocket interface so Stream Deck dials can be set up to control station volume.

With this in mind, is the current state of a station's volume being maintained in the correct place after this PR? Looks like it is in the backend. Should it instead be even deeper and directly in afv-native (e.g. so CRC can make use of this too)?

GeorgeBarlow commented:
> With this in mind, is the current state of a station's volume being maintained in the correct place after this PR? Looks like it is in the backend. Should it instead be even deeper and directly in afv-native (e.g. so CRC can make use of this too)?

The state is already maintained within afv-native inside `AtcRadioState`

GeorgeBarlow commented:
Plan of action: PR will be merged into a branch called ws-gain, which is overhauling the volume system. We will use the front end code and the C++ code to help with what we are doing in that branch (which is aiming to centralise and add mute functionality to radios)

GeorgeBarlow commented:
Merged! Thank you for contributing a great feature, your commits will still get shown as authored by you and recognised in the release notes.
